src/qt/walletmodel.cpp

```diff
@@ -120,12 +120,23 @@ void WalletModel::pollBalanceChanged()
 
 void WalletModel::checkBalanceChanged(const interfaces::WalletBalances& new_balances)
 {
-    if(new_balances.balanceChanged(m_cached_balances)) {
-        m_cached_balances = new_balances;
+    const auto& cached_balace = getCachedBalance();
+    if (new_balances.balanceChanged(cached_balace)) {
+        {
+            // Only place where the balance cache is updated.
+            QMutexLocker locker(&m_cache_mutex);
+            m_cached_balances = new_balances;
+        }
         Q_EMIT balanceChanged(new_balances);
     }
 }
 
+interfaces::WalletBalances WalletModel::getCachedBalance() const
+{
+    QMutexLocker locker(&m_cache_mutex);
```

Comment:
0ac7d0cd538d79511105e0e4f69dc6e2a8f10a44

Are there concurrent accesses?

Reply:
Hey, not yet. It's a glance into the future actually hehe.

At the moment it's only accessed from the main thread. At the following locations:

1) From `pollBalanceChanged` (`QTimer`).
2) From `setWalletModel` -> in `OverviewPage` and `SendCoinsDialog`.
3) From `useAvailableBalance` in `SendCoinsDialog`.
4) From `prepareTransaction` in `SendCoinsDialog`.

Now the rationale:
 I'm locally working on a branch that moves part of the processes to a worker thread thus why the guard. And added that commit here mainly because it would let me work on different flows in different scoped PRs without having to add it to each of them. 
- Not a big deal for me to remove it from this one -

Reply:
Personally, a good rule to follow is to only push stuff that is used or necessary.

Reply:
np 👍🏼, dropped.